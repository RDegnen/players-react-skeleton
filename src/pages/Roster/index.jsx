import React, { Fragment } from 'react';
import { Button, Table } from 'reactstrap';

import constants from '../../constants';

const columns = [
  {
    header: 'First Name',
    value: ({ first_name: firstName }) => firstName,
  },
  {
    header: 'Last Name',
    value: ({ last_name: lastName }) => lastName,
  },
  {
    header: 'Handedness',
    value: ({ handedness }) => handedness,
  },
  {
    header: 'Rating',
    value: ({ rating }) => rating,
  },
];

class Roster extends React.Component {
  constructor(props) {
    super(props);
    this.state = { players: [] };
    this.getRoster = this.getRoster.bind(this);
    this.deletePlayer = this.deletePlayer.bind(this);
  }

  componentDidMount() {
    this.getRoster();
  }

  getRoster() {
    const authToken = localStorage.getItem('token');
    return fetch(`${constants.API_URL}/api/players`, {
      headers: {
        Authorization: `Bearer ${authToken}`,
      },
    }).then(res => res.json())
      .then(json => this.setState({ players: [...json.players] }))
      .catch(err => err);
  }

  deletePlayer(playerId) {
    const authToken = localStorage.getItem('token');
    return fetch(`${constants.API_URL}/api/players/${playerId}`, {
      method: 'DELETE',
      headers: {
        Authorization: `Bearer ${authToken}`,
      },
    }).then(res => res.json())
      .then(() => this.getRoster())
      .catch(err => err);
  }

  render() {
    const { players } = this.state;
    return (
      <Fragment>
        <h2>Roster</h2>
        <Table responsive>
          <thead>
            <tr>
              {columns.map(({ header }) => (
                <th key={header}>
                  {header}
                </th>
              ))}
            </tr>
          </thead>
          {players.map(player => (
            <tr key={player.id}>
              {columns.map(({ header, value }) => (
                <td key={header}>{value(player)}</td>
              ))}
              <td key={`${player.id}Delete`}>
                <Button className="delete" onClick={() => this.deletePlayer(player.id)}>Delete</Button>
              </td>
            </tr>
          ))}
        </Table>
      </Fragment>
    );
  }
}

export default Roster;
