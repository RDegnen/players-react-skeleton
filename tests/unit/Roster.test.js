import React from 'react';
import { shallow } from 'enzyme';

import Roster from '../../src/pages/Roster';

global.fetch = jest.fn(() => new Promise(resolve => resolve()));
const getComponent = props => shallow(<Roster {...props} />);

describe('<Roster />', () => {
  it('should match snapshot', () => {
    const component = getComponent();
    expect(component).toMatchSnapshot();
  });

  it('should render a player roster when the state is updated', () => {
    const component = getComponent();
    const players = [
      {
        first_name: 'Randy',
        last_name: 'Marsh',
        handedness: 'right',
        rating: 50,
        id: 1,
      },
    ];
    component.instance().setState({ players });
    expect(component.find('td').length).toBe(5);
    expect(component.find('td').first().text()).toBe('Randy');
  });
});
