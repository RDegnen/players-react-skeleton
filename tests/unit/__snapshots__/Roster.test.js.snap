// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Roster /> should match snapshot 1`] = `
<Fragment>
  <h2>
    Roster
  </h2>
  <Table
    responsive={true}
    responsiveTag="div"
    tag="table"
  >
    <thead>
      <tr>
        <th
          key="First Name"
        >
          First Name
        </th>
        <th
          key="Last Name"
        >
          Last Name
        </th>
        <th
          key="Handedness"
        >
          Handedness
        </th>
        <th
          key="Rating"
        >
          Rating
        </th>
      </tr>
    </thead>
  </Table>
</Fragment>
`;
